docs/historical_specification.rst

```diff
@@ -0,0 +1,668 @@
+reduction have to be applied to Intersections first:
+
+- Nested intersections shall be flattened, i.e ``Intersection[A, Intersection[B, C]] ==
+  Intersection[A, B, C]``
+- If a (concrete or protocol) type ``A`` is a subtype of ``B``, ``A`` shall be removed from the
+  intersection
+- If a protocol ``BP`` defines **all** methods and properties of a protocol ``AP``, ``AP`` shall be
+  removed from the intersection
+- If the concrete class ``A`` fulfils the Protocol ``AP``, ``AP`` shall be removed from the
+  intersection
+- An intersection with only one element shall be normalized to the element.
+
+
+``Any`` Reduction
+-----------------
+As `PEP 483`_ already suggested: ``Any`` shall be removed from an ``Intersection``, i.e.
+``Intersection[A, B, Any] == Intersection[A, B]``.
+
+% This is only a suggestion and needs to be discussed and decided in https://github.com/CarliJoy/intersection_examples/issues/1
+% Once it was finally decided the discussion and arguments should be summarized here.
+
+
+``Never`` Evaluation
+--------------------
+An intersection that contains either two classes that are a or are a subclass of two different `internal base classes <https://docs.python.org/3/library/stdtypes.html>`_ shall evaluate to ``Never``.
+Examples for internal baseclasses are:
+
+- BaseException
+- bool
+- bytearray
+- bytes
+- complex
+- dict
+- float
+- frozenset
+- int
+- list
+- memoryview
+- range
+- set
+- str
+- tuple
+- type
+
+There are concrete types that can't be subclassed, they are
+ - a class marked with ``typing.final`` `[doc] <https://docs.python.org/3/library/typing.html#typing.final>`_
+ - ``typing.Never`` and ``typing.NoReturn`` also called `bottom type <https://en.wikipedia.org/wiki/Bottom_type>`_
+ - ``None``
+
+If such a type is used within an intersection this intersection shall evaluate to ``Never``.
+
+The reasoning behind this is that these types can't be subtyped and shouldn't be dynamically
+extended.
+Doing this early prevents issues during subtyping or assignments checks.
+
+::
+
+    from typing import TypeVar, reveal_type
+
+    T = TypeVar("T")
+
+    class Enhanced:
+        is_great: bool
+
+
+    def enhance(cls: type[T]) -> type[T & Enhanced]:
+        class New(cls, Enhanced):
+            ...
+
+        return New
+
+    reveal_type(enhance(str))  # okay
+    reveal_type(enhance(None))  # raises a TypeError on runtime, should be flagged by TypeCheckers
+
+It is important to note that once a type checker evaluated anything to ``Never`` within an
+intersection it can stop further evaluations an return ``Never``.
+This way a lot of edge cases by mixin types that can't be mixed are handled easily.
+
+Handling Callables
+------------------
+Every Callable within an intersection shall be treated like a ``def __call__()`` Protocol.
+
+::
+
+    from typing import Protocol, Callable
+
+    MyCallable = Callable[[str, int], float]
+
+    class CallProto:
+        def __call__(a: str, b: int) -> float: ...
```

Comment:
```suggestion
        def __call__(a: str, b: int, /) -> float: ...
```

Might be good to make the callable protocol positional-only; with normal arguments `a` and `b` the types aren't exactly equivalent.